File: 1_1_showdown/internal/domain/deck.go
package domain

import "fmt"

type Deck struct {
	cards []Card
}

var (
	Ranks = []int{
		RANK2, RANK3, RANK4, RANK5,
		RANK6, RANK7, RANK8, RANK9,
		RANK10, RANKJ, RANKQ, RANKK, RANKA,
	}
	Suits = []int{
		SuitClub, SuitDiamond, SuitHeart, SuitSpade,
	}
)

func NewDeck() Deck {
	cards := make([]Card, 0, 52)

	for _, r := range Ranks {
		for _, s := range Suits {
			cards = append(cards, Card{rank: r, suit: s})
		}
	}

	return Deck{
		cards: cards,
	}
}

func (d *Deck) DrawCard() Card {
	c := d.cards[0] // pop first one
	d.cards = d.cards[1:]
	return c
}

func (d *Deck) Shuffle() {
	// do nothing for first version
	fmt.Println("Shuffle")
}

func (d *Deck) Size() int {
	return len(d.cards)
}

File: 1_1_cpp/domain/deck.cpp

#include "deck.h"
#include <iostream>

const std::vector<int> Deck::Ranks = {
    Card::RANK2, Card::RANK3, Card::RANK4, Card::RANK5,
    Card::RANK6, Card::RANK7, Card::RANK8, Card::RANK9,
    Card::RANK10, Card::RANKJ, Card::RANKQ, Card::RANKK, Card::RANKA
};

const std::vector<int> Deck::Suits = {
    Card::SuitClub, Card::SuitDiamond, Card::SuitHeart, Card::SuitSpade
};

Deck::Deck() {
    cards.reserve(52);

    for (int r : Ranks) {
        for (int s : Suits) {
            cards.push_back(Card(r, s));
        }
    }
}

Card Deck::DrawCard() {
    Card c = cards[0]; // pop first one
    cards.erase(cards.begin());
    return c;
}

void Deck::Shuffle() {
    // do nothing for first version
    std::cout << "Shuffle" << std::endl;
}

int Deck::Size() const {
    return static_cast<int>(cards.size());
}